src/scratchOrgInfoApi.ts

```diff
@@ -0,0 +1,504 @@
+      if (resultInProgress.Status === 'Active' || resultInProgress.Status === 'Error') {
+        return resultInProgress;
+      }
+      // all other statuses, OR lack of status (e.g. network errors) will cause a retry
+      throw new Error(`Scratch org status is ${resultInProgress.Status}`);
+    },
+    {
+      retries: 'INFINITELY',
+      timeout: timeout.milliseconds,
+      delay: Duration.seconds(2).milliseconds,
+      backoff: 'LINEAR',
+      maxBackOff: Duration.seconds(30).milliseconds,
+    }
+  ).catch(() => {
+    throw new SfdxError(`The scratch org did not complete within ${timeout.minutes} minutes`, 'orgCreationTimeout', [
+      'Try your force:org:create command again with a longer --wait value',
+    ]);
+  });
+  return checkScratchOrgInfoForErrors(response, hubOrg.getUsername(), logger);
+};
+
+/**
+ * This authenticates into the newly created org and sets org preferences
+ *
+ * @param scratchOrgAuthInfo - an object containing the AuthInfo of the ScratchOrg
+ * @param apiVersion - the target api version
+ * @param orgSettings - The ScratchOrg settings
+ * @param scratchOrg - The scratchOrg Org info
+ * @returns {Promise<Optional<AuthInfo>>}
+ */
+export const deploySettingsAndResolveUrl = async (
+  scratchOrgAuthInfo: AuthInfo,
+  apiVersion: string,
+  orgSettings: SettingsGenerator,
+  scratchOrg: Org
+): Promise<Optional<AuthInfo>> => {
+  const logger = await Logger.child('scratchOrgInfoApi-deploySettingsAndResolveUrl');
+
+  if (orgSettings.hasSettings()) {
+    // deploy the settings to the newly created scratch org
+    logger.debug(`deploying scratch org settings with apiVersion ${apiVersion}`);
+
+    try {
+      await orgSettings.createDeploy();
+      await orgSettings.deploySettingsViaFolder(scratchOrg, apiVersion);
+    } catch (error) {
+      throw SfdxError.wrap(error as Error);
+    }
+  }
+
+  if (scratchOrgAuthInfo.getFields().instanceUrl) {
+    logger.debug(
+      `processScratchOrgInfoResult - resultData.instanceUrl: ${JSON.stringify(
+        scratchOrgAuthInfo.getFields().instanceUrl
+      )}`
+    );
+    const options = {
+      timeout: Duration.minutes(3),
+      frequency: Duration.seconds(10),
+      url: new SfdcUrl(ensureString(scratchOrgAuthInfo.getFields().instanceUrl)),
+    };
+    try {
+      const resolver = await MyDomainResolver.create(options);
+      await resolver.resolve();
+    } catch (error) {
+      const sfdxError = SfdxError.wrap(error as Error);
+      logger.debug(`processScratchOrgInfoResult - err: ${JSON.stringify(error, null, 4)}`);
+      if (sfdxError.name === 'MyDomainResolverTimeoutError') {
+        sfdxError.setData({
+          orgId: scratchOrgAuthInfo.getFields().orgId,
+          username: scratchOrgAuthInfo.getFields().username,
+          instanceUrl: scratchOrgAuthInfo.getFields().instanceUrl,
+        });
+        logger.debug(`processScratchOrgInfoResult - err data: ${JSON.stringify(sfdxError.data, null, 4)}`);
```

Comment:
fyi: you can change this to be `'- err data: %s', sfdxError.data)` and it will handle formatting an object for you